### PR TITLE
Clojure API

### LINKED ISSUE #55: Clojure API
In a similar vein to https://github.com/clj-kondo/clj-kondo/issues/226, Carve could offer a clojure API.

Checking out https://github.com/borkdude/carve/blob/f499f65e36e02484609f01ce891f3c0207b24444/src/carve/main.clj#L64-L80, a few things stand out as not very api-ish:

* a `println`
* CLI option parsing
* Printing of reports (even when requesting edn) instead of returning edn directly

WDYT?

Thanks - V

### PR DESCRIPTION
**Pull request by borkdude**
Fixes #55

### CODE REVIEW

**Comment by borkdude** [comment]
@vemv I made a different approach here. The `run!` function in `carve.api` works similar to the CLI api, but the `report` function works better for "just data". Also, similar to the clj-kondo API, there is a `print!` function that can be passed a result from `run!` for printing.

```
$ clj
Clojure 1.10.1
user=> (require '[carve.api :as api])
nil
user=> (api/report {:paths ["src"]})
{:report {:unused-vars ({:filename "src/carve/api.clj", :row 5, :col 1, :ns carve.api, :name print!} {:filename "src/carve/api.clj", :row 9, :col 1, :ns carve.api, :name run!} {:filename "src/carve/api.clj", :row 19, :col 1, :ns carve.api, :name report} {:filename "src/carve/impl.clj", :row 343, :col 1, :ns carve.impl, :name run!})}, :config {:paths ["src"], :report true}}
user=> (api/print! (api/run! {:paths ["src"] :report {:format :text}}))
src/carve/api.clj:5:1 carve.api/print!
src/carve/api.clj:9:1 carve.api/run!
src/carve/api.clj:19:1 carve.api/report
src/carve/impl.clj:343:1 carve.impl/run!
```

**Comment by borkdude** [comment]
@vemv I may make some tweaks to this, but feel free to try it out.

File: src/carve/api.clj [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+  CLI."
+  ([] (run! nil))
+  ([opts]
+   (impl/run+ opts)))
+
+(defn report
+  "Similar to run! but forces non-side-effecting non-interactive behavior."
+  ([] (report {:merge-config true}))
+  ([opts]
+   (impl/run+ (assoc opts :report true))))
```
Comment (vemv) [inline src/carve/api.clj:21]: WDYT about passing `:silent true` as well? Currently it doesn't make a difference, but that seems subject to change e.g. you might add logging in a new place tomorrow.

File: src/carve/impl.clj [hunk 2]
```diff
@@ -271,3 +273,74 @@
                          (dissoc analysis :unused-var-refers)))
                 (reportize results)))
           (reportize results))))))
+
+(set! *warn-on-reflection* true)
+(s/check-asserts true)
```
Comment (vemv) [inline src/carve/impl.clj:278]: Not sure if `check-asserts` will be limited to the current ns?

**Review by vemv — COMMENTED** [review]
Tried out, works for me! Thanks.

Perhaps you could salvage some tests from https://github.com/borkdude/carve/pull/57/files - I intend to give good use to the API so ideally I'd like it to be well supported

File: src/carve/api.clj (hunk 1 above)
Comment (borkdude) [reply src/carve/api.clj:21]: `:silent` is about writing stuff to stdout for interactivity. You can rely on `:report` always implying `:silent true`.

Logging warnings (like in the main) may happen to warn users of important stuff like breakage or something, despite this setting. In that case you can rebind `*err*` to a new StringWriter to ignore the output, if that should happen.

File: src/carve/impl.clj (hunk 2 above)
Comment (borkdude) [reply src/carve/impl.clj:278]: This is a global flag, but good point. I think we should only set this in the main ns (which is intended to be used from the CLI / native). Setting this as part of a library isn't a good idea, this choice should be done by the user.

**Comment by borkdude** [comment]
@vemv I'm just going to merge this and you can follow up with incremental changes, like tests, if desired. If you choose not to, I will copy (some of) your tests from the other PR.

**Pull request merged**